keps/sig-apps/20190523-support-completions-index-for-job.md

```diff
@@ -0,0 +1,78 @@
+---
+title: support-completions-index-for-job
+authors:
+  - "@fudali"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-job
+reviewers:
+  - "@kow3ns"
+  - "@janetkuo"
+approvers:
+  - "@kow3ns"
+  - "@janetkuo"
+editor: TBD
+creation-date: 2019-05-23
+last-updated: 2019-05-27
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# support completions index for JOB
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+
+## Summary
+
+[Job](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/)
+is a Kubernetes CRD that help us run task, Job allows a user to specify the task or parallel task through either `podTemplate` and `completions`; but i use parallel task i must ensure task unrepeat in myself application, this is diffcult; if job support sahrds, i can use shards index to implement task unrepeat is so easy;
+
+## Motivation
+
+I have a case need shards to job, i want to use kubernetes implement my case; i read the [document](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#parallel-jobs), this is a not implement function, so i want implement it;
+
+### Goals
+
+- Implement each Pod is passed a different index in the range 1 to .spec.completions.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+In the current implementation, the Job controller not allocation index on `.completions != nil`; in my implementation, when `.completions != nil` i will through completions value and succeeded pods get the available indexes, i will record index to the job create pods, and set env to the pod （user application get index is easy）; next i will allocation index to creating pod function, i will lock this step ensure not repeat index not more than one, if one index repeat, i will kill one on next this job created pod changes time;
```

Comment:
How do we persist the index, e.g. annotation, new field of Pod? If not persist, how to recover it when kube-controller-manager failover or change leader? 

BTW, please also highlight the interface in design doc for review, e.g. env name ?

Reply:
ok, i will fix it;

i persist the index to labels, i choice label because i think easy to use on `kubectl get po -l`, label name is `job-completions-index` and i set `JOB_COMPLETIONS_INDEX ` env in pod (user get this index in application easier to use )